### ISSUE TITLE
Trouble with cuda out of memory

### ISSUE THREAD

**Issue by EduardoLawson1** [google-research/semivl#10]
Hi guys! Incredible work! I'm trying to train the model on the pascal voc with 92 labels on a NVIDIA TITAN X gpu. Unfortunately I keep getting the following error:  ` RuntimeError: CUDA out of memory.Tried to allocate 98.00 MiB (GPU 0; 11.90 GiB total capacity; 10.09 GiB already allocated; 100.38 MiB free; 10.42 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF `. 
So I did some research in how to use this function for cuda allocation and tried the following with some different values. 
`os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "max_split_size_mb:128"` I added this line of code in the main of experiments.py but still doesn't work. My question is if I'm doing something wrong or my gpu just don't have enough memory to run this model.

**Comment by EduardoLawson1** [comment]
Thank you for the advices! I will give it a try and if it works I will let you know.

**Comment by joshuakurien** [comment]
Hi,

I'm trying to run the training for the model on 2 x A6000 GPUs and am still running into similar issues. I'm getting the following error: 

`RuntimeError: CUDA out of memory. Tried to allocate 388.00 MiB (GPU 1; 47.53 GiB total capacity; 44.65 GiB already allocated; 83.31 MiB free; 45.02 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF`

Any idea what I could do to fix this? 

Thanks!

**Comment by haruishi43** [comment]
Hey @lhoyer, I managed to train the code using two A6000 GPUs by making a few changes to the code. Here’s what I did:

1. I set `need_weights=False` when passing `nn.MultiheadAttention`. This uses the optimized `scaled_dot_product_attention`, which saves a lot of memory (torch > 2.0).
2. I used mixed precision with half-precision.

With these changes, I was able to reduce memory usage from 21.6GB to 14.9GB when using 2 labeled and 2 unlabeled images with a crop size of 512x512.

When training with 8 labeled and 8 unlabeled images, I needed 2 A6000 GPUs with 4 labeled and 4 unlabeled images for each GPU. Each GPU used around 26GB of memory.

For Cityscapes, it still requires 4 A100 GPUs with 8 labeled and 8 unlabeled images.

I also tried gradient checkpointing by modifying `TransformerEncoderLayer`, but I got some errors.
